### ISSUE TITLE
Support Sqlite

### ISSUE THREAD

**Issue by bouwe77** [bouwe77/temba#24]
# TL;DR

What: Support Sqlite as an alternative to in-memory and MongoDB storage
Why: So we don't need a database server

# Thoughts

The only database Temba supports is MongoDB. This is a nice fit, because it allows storing JSON blobs, which can be queried easily, might Temba support filtering in the future.

However, MongoDB requires a server, so additional support for Sqlite, where databases are just files, would be a nice alternative. The simplicity of Sqlite would fit Temba's simplicity, as well the wide support for hosting Sqlite by modern cloud platforms.

Also, Temba's in memory persistence could then also be replaced by Sqlite, as that supports in-memory databases as well. That would mean Temba would always store data in Sqlite, unless you configure a Mongo DB connection string.

Sqlite, however, is a relational database, not a JSON document store, so we can not store JSON documents that are filterable. In our Sqlite database we would create a table for each resource, and within each table have `id` and `json` fields. Because the `json` field contains a JSON blob, filtering would be really inefficient, as you need to read all data of the whole resource table into memory, parse the JSON, and only then would be able to do filtering.

An alternative would be to dynamically create fields in a resource table for each JSON field. However, this would become complex to parse, and we would loose quite a lot of flexibility and simplicity Temba has now.

Another approach would be any other solution that has file based storage of JSON documents, offering similar functionality as MongoDB. However, hosting platforms should also support this. For example, when you use json-server, which stores JSON files, hosting platforms might not fully support this. I had a situation where Heroku deleted JSON files because of how their file storage was set up. That is a benefit of Sqlite, which is very well supported by all modern cloud hosting platforms.

So with all this being said, Sqlite is pretty low on the priority list of possible Temba features...

**Comment by bouwe77** [comment]
Closing as we won't fix this
